nixos/lib/test-driver/lib/machine.py

```diff
@@ -0,0 +1,794 @@
+    "Z": "shift-z",
+    "\n": "ret",
+    "!": "shift-0x02",
+    "@": "shift-0x03",
+    "#": "shift-0x04",
+    "$": "shift-0x05",
+    "%": "shift-0x06",
+    "^": "shift-0x07",
+    "&": "shift-0x08",
+    "*": "shift-0x09",
+    "(": "shift-0x0A",
+    ")": "shift-0x0B",
+}
+
+
+def make_command(args: list) -> str:
+    return " ".join(map(shlex.quote, (map(str, args))))
+
+
+def retry(fn: Callable, timeout: int = 900) -> None:
+    """Call the given function repeatedly, with 1 second intervals,
+    until it returns True or a timeout is reached.
+    """
+
+    for _ in range(timeout):
+        if fn(False):
+            return
+        time.sleep(1)
+
+    if not fn(True):
+        raise Exception(f"action timed out after {timeout} seconds")
+
+
+def _perform_ocr_on_screenshot(
+    screenshot_path: str, model_ids: Iterable[int]
+) -> List[str]:
+    if shutil.which("tesseract") is None:
+        raise Exception("OCR requested but enableOCR is false")
+
+    magick_args = (
+        "-filter Catrom -density 72 -resample 300 "
+        + "-contrast -normalize -despeckle -type grayscale "
+        + "-sharpen 1 -posterize 3 -negate -gamma 100 "
+        + "-blur 1x65535"
+    )
+
+    tess_args = f"-c debug_file=/dev/null --psm 11"
+
+    cmd = f"convert {magick_args} {screenshot_path} tiff:{screenshot_path}.tiff"
+    ret = subprocess.run(cmd, shell=True, capture_output=True)
+    if ret.returncode != 0:
+        raise Exception(f"TIFF conversion failed with exit code {ret.returncode}")
+
+    model_results = []
+    for model_id in model_ids:
+        cmd = f"tesseract {screenshot_path}.tiff - {tess_args} --oem {model_id}"
+        ret = subprocess.run(cmd, shell=True, capture_output=True)
+        if ret.returncode != 0:
+            raise Exception(f"OCR failed with exit code {ret.returncode}")
+        model_results.append(ret.stdout.decode("utf-8"))
+
+    return model_results
+
+
+class Machine:
+    def __init__(self, args: Dict[str, Any]) -> None:
+        self.name = args.get("name") or "machine"
+        self.log_serial = args["log_serial"]
+        self.log_machinestate = args["log_machinestate"]
+        self.script = args.get("startCommand", self.create_startcommand(args))
+
+        # in order to both enable plain log functions, but also nested
+        # machine state logging, check if the `log_machinestate` object
+        # has a method `nested` and provide this functionality.
+        # Ideally, just remove this logic long term and make it as  easy as
+        # possible for the user of `Machine` to provide plain standard loggers
+        @contextmanager
+        def dummy_nest(message: str) -> _GeneratorContextManager:
+            self.log_machinestate(message)
+            yield
+
+        nest_op = getattr(self.log_machinestate, "nested", None)
+        if callable(nest_op):
+            self.nested = nest_op
+        else:
+            self.nested = dummy_nest
+
+        # Can be used in child classes
```

Comment:
The goal of `machine_config_modifier` was to not subclass the `Machine` class. If you look at the tmux driver implementation, we never inherit from the `Machine` class. (Few months ago, we had a  long discussion on that topic).

Reply:
I like the callback more even if it's probanly "unpythonic". It's pynix in a sense. :smile:

Reply:
This is now currently gone. We'll see how the tmux driver fits in, when we come to it.